validphys2/src/validphys/commondataparser.py

```diff
@@ -1,18 +1,319 @@
+    conversion_factor: float = 1.0 
+    comment: Optional[str] = None
+    apfelcomb: Optional[ValidApfelComb] = None
+    # The following options are transitional so that the yamldb can be used from the theory
+    appl: Optional[bool] = False
+    target_dataset: Optional[str] = None
+
+    def fktables_to_paths(self, grids_folder):
+        """Given a source for pineappl grids, constructs the lists of fktables
+        to be loaded"""
+        ret = []
+        for operand in self.FK_tables:
+            ret.append([grids_folder / f"{m}.{EXT}" for m in operand])
+        return ret
+    
+    @classmethod
+    def parser(cls, meta: dict):
+        """The yaml databases in the server use "operands" instead of "FK_tables" """
+        if "operands" in meta:
+            meta["FK_tables"] = meta.pop("operands")
+        return parse_input(meta, cls)
+
+
+@dataclass
+class KinematicsMeta:
+    """Contains all metadata for the kinematics of the dataset"""
+
+    file: ValidPath
+    variables: dict
+
+    @classmethod
+    def parser(cls, meta: dict):
+        return parse_input(meta, cls)
+
+
+@dataclass
+class ReferenceMeta:
+    """Holds literature information for the dataset"""
+
+    url: str
+    version: int = 0
+    tables: list[int] = field(default_factory=list)
+
+    @classmethod
+    def parser(cls, meta: dict):
+        return parse_input(meta, cls)
+
+
+@dataclass
+class Variant:
+    """Defines the keys of the CommonMetaData that can be overwritten"""
+
+    data_uncertainties: list[ValidPath]
+
+
+# Define parsers for the more complicated structures
+ValidTheory = Parser(TheoryMeta.parser)
+ValidReference = Parser(ReferenceMeta.parser)
+ValidKinematics = Parser(KinematicsMeta.parser)
+
+
+@Parser
+def ValidVariants(variant_dict: dict) -> dict:
+    """Variants of a dataset are allowed to overwrite a subset of the keys of a dataset
+    (those defined in the Variant dataclass).
+    This wrapper class runs over the dictionary of variant and parses them into valid Variants
+    """
+    return {k: parse_input(v, Variant) for k, v in variant_dict.items()}
+
+
+@dataclass
+class CommonMetaData:
+    setname: str
+    ndata: int
+    observable: dict
+    kinematics: ValidKinematics
+    kinematic_coverage: dict
+    data_central: ValidPath
+    data_uncertainties: list[ValidPath]
+    dataset_label: str
+    plot_x: str
+    figure_by: list[str]
+    theory: ValidTheory
+    nnpdf_metadata: dict
+    version: int
+    version_comment: str = ""
+    arXiv: Optional[ValidReference] = None
+    iNSPIRE: Optional[ValidReference] = None
+    hepdata: Optional[ValidReference] = None
+    variants: Optional[ValidVariants] = None
+
+    _enabled_variant: list[str] = field(default=None, repr=False)
+    _default: dict = field(default=None, repr=False)
+
+    # TODO: these methods will be moved to the main CommonData class
+    # as any change of variant should trigger a reload of the commondata
+    def enable_variant(self, variant):
```

Comment:
I think we should have a class CommonMetaData without any mutable state and Commondata(common_meeta_data, variant) also without mutable state. I don;t know if that is what the comment above is about.

Reply:
No it is not, but it makes sense, yes.